tools/Makefile.shared

```diff
@@ -12,10 +12,62 @@
 #*   special exception on linking described in the file LICENSE.          *
 #*                                                                        *
 #**************************************************************************
-
+MAKEFLAGS := -r -R
 include ../config/Makefile
-CAMLRUN ?= ../boot/ocamlrun
-CAMLYACC ?= ../boot/ocamlyacc
+ifndef CAMLRUN
+CAMLRUN := ../boot/ocamlrun
+endif
+ifndef CAMLYACC
+CAMLYACC := ../boot/ocamlyacc
+endif
+DESTDIR ?=
+
+# Setup GNU make variables storing per-target source and target,
+# a list of installed tools, and a function to quote a filename for
+# the shell.
+override target=''""'$(subst ','\'',$@)'#'#')'#
+override source=''""'$(subst ','\'',$<)'#'#')'#
```

Comment:
This is really overkill. We don't have source (or target) files with special characters in them, and `ocamlc` would warn about them anyway. Moreover, you're also checking for the absence of spaces in file names.